### PR TITLE
[Bug Fix]: Update customer and vendor joins in transaction details

### PR DESCRIPTION
**Pull request by fivetran-avinash**
## PR Overview
**This PR will address the following Issue/Feature:** #109 

**This PR will result in the following new package version:** v0.14.0

There are some fields added and underlying logic changed. It isn't necessarily a breaking change but will be part of a wider release with other issues.

**Please provide the finalized CHANGELOG entry which details the relevant changes included in this PR:**

## Bug Fixes
- Updates logic in `netsuite2__transaction_details` to select the appropriate customer and vendor values based on the whether the transaction type is a customer invoice or credit, or a vendor bill or credit.

## Feature Updates
- New fields `customer_alt_name` and `vendor_alt_name` were introduced into `netsuite2__transaction_details`, after being added into the `stg_netsuite2__customers` and `stg_netsuite2__vendors` models in the most recent release of `dbt_netsuite`. [You can view the release notes](https://github.com/fivetran/dbt_netsuite/releases/tag/v0.14.0) for more details.

## Under the Hood
- Generated consistency test for `netsuite2__transaction_details` to make sure amounts match for integration test validations. 

## Contributors
- [@jmongerlyra](https://github.com/jmongerlyra) ([PR #131](https://github.com/fivetran/dbt_netsuite/pull/131))

### Basic Validation
Please acknowledge that you have successfully performed the following commands locally:
- [x] dbt run –full-refresh && dbt test
- [x] dbt run (if incremental models are present) && dbt test

Before marking this PR as "ready for review" the following have been applied:
- [x] The appropriate issue has been linked, tagged, and properly assigned
- [x] All necessary documentation and version upgrades have been applied
- [x] docs were regenerated (unless this PR does not include any code or yml updates)
- [x] BuildKite integration tests are passing
- [x] Detailed validation steps have been provided below

### Detailed Validation
Please share any and all of your validation steps:

New consistency test to check amounts in transaction details passed.
Screenshot 2024-09-16 at 8 38 56 AM

### If you had to summarize this PR in an emoji, which would it be?
 
🩺

### CODE REVIEW

File: integration_tests/tests/consistency/consistency_transaction_details_amounts.sql
Comment (fivetran-joemarkiewicz) [inline integration_tests/tests/consistency/consistency_transaction_details_amounts.sql:36]: I know this test is passing. However, there are scenarios where if a transaction_id is not present in dev then we will just see a `null` populated. Then a valid number - `null` just returns `null` and essentially passes the where condition, but in reality there is a mismatch in the data. 

image

See below for a suggestion to sum by the transaction date so we can look at the aggregate sum of the converted and transaction amounts. This way if there is a `null` in dev but not in prod then we should see less in the aggregate and therefore a resulting failure in the validation test. Additionally, I included a row count in here as well for extra validation. When testing locally I could see that the row counts matched between dev and prod confirming there is no fan out resulting from the new joins.

```suggestion
with prod as (

    select 
        cast(transaction_date as date) as date_day,
        count(*) as prod_row_count,
        sum(converted_amount) as prod_converted_amount,
        sum(transaction_amount) as prod_transaction_amount
    from {{ target.schema }}_netsuite_prod.netsuite2__transaction_details
    group by 1
),

dev as (

    select 
        cast(transaction_date as date) as date_day,
        count(*) as dev_row_count,
        sum(converted_amount) as dev_converted_amount,
        sum(transaction_amount) as dev_transaction_amount
    from {{ target.schema }}_netsuite_dev.netsuite2__transaction_details
    group by 1

),

final as (

    select
        prod.date_day,
        prod.prod_row_count,
        dev.dev_row_count,
        prod.prod_converted_amount,
        dev.dev_converted_amount,
        prod.prod_transaction_amount,
        dev.dev_transaction_amount
    from prod
    full outer join dev
        on dev.date_day = prod.date_day
)
```

File: integration_tests/tests/consistency/consistency_transaction_details_amounts.sql
Comment (fivetran-joemarkiewicz) [inline integration_tests/tests/consistency/consistency_transaction_details_amounts.sql:41]: Added condition with the new aggregate suggested.
```suggestion
or abs(prod_converted_amount - dev_converted_amount) >= 0.01
or abs(prod_row_count - dev_row_count) >= 0.01
```

File: packages.yml
Comment (fivetran-joemarkiewicz) [inline packages.yml:7]: Reminder to update before merge/release

File: CHANGELOG.md
Comment (fivetran-joemarkiewicz) [inline CHANGELOG.md:5]: Can you include all the previous fields which are impacted from this code fix.

**Review by fivetran-joemarkiewicz — CHANGES_REQUESTED** [review]
@fivetran-avinash thanks for working through these changes! They are looking good to me and I have no concerns regarding the resulting data. 

I do have a few quick requests and suggestions before approval. Let me know if you have any questions. Thanks!

File: README.md
Comment (fivetran-joemarkiewicz) [inline README.md:141]: Don't forget to include the employee disable/enable documentation in the README as well since that is an upstream change.

File: CHANGELOG.md
Comment (fivetran-joemarkiewicz) [inline CHANGELOG.md:1]: Please also include the relevant upstream breaking and other important changes in this changelog entry

**Review by fivetran-joemarkiewicz — COMMENTED** [review]
@fivetran-avinash please see a few additional notes I had following my re-review of the source PR.

Additionally, I may be missing something, but do we reference the new employee staging table in this PR at all? Or is the intention that the new employee staging model will be referenced in a different PR? Also, with the new enable/disable variable from the upstream dependency, we will need to add that to the [quickstart.yml](https://github.com/fivetran/dbt_netsuite/blob/main/.quickstart/quickstart.yml).

File: README.md
Comment (fivetran-avinash) [reply README.md:141]: Brought it in!

File: CHANGELOG.md
Comment (fivetran-avinash) [reply CHANGELOG.md:1]: Added.

**Review by fivetran-avinash — COMMENTED** [review]
@fivetran-joemarkiewicz This is ready for re-review. The employee staging model nor any of its fields are being brought downstream to the transformations model. I made a note about this in the CHANGELOG. 

I also slightly tweaked the consistency test, but that is probably because I ran my test very late as transactions are still coming in from the previous day. Open to removing the additional filter I put in.

**Review by fivetran-joemarkiewicz — CHANGES_REQUESTED** [review]
@fivetran-avinash this looks great! Thanks for making the final changes. The very last item needed before approval is to update the quickstart.yml to include the employee disable situation.

Also I want to confirm with our internal team that the employee staging model not being used in a downstream model won't be an issue.

File: integration_tests/tests/consistency/consistency_transaction_details_amounts.sql
Comment (fivetran-avinash) [reply integration_tests/tests/consistency/consistency_transaction_details_amounts.sql]: Updated. Made some slight changes (the most recent day can see late arriving transactions when I did a devprod run and lead to some slight discrepancies on transactions) so I created an additional filter here to get the consistency test to pass.

**Review by fivetran-avinash — COMMENTED** [review]
@fivetran-joemarkiewicz Final PR updates made!

**Review by fivetran-joemarkiewicz — APPROVED** [review]
@fivetran-avinash these changes look good to me! This should be good to merge into a release branch and incorporate the other changes for Netsuite.

**Pull request merged**
